rockchip: rk3399 fix PCIE reboot issues

wurzerj opened the pull request:
Fix LAN port failing to initialize after reboot particularly on some FriendlyElec NanoPi R4S devices. Interestingly this error does not seem to occur on all devices (independent of whether it is an Enterprise edition or a Standard edition). Maybe it is related to a hardware revision change ... just guessing.

The patch itself is forked from Armbian. Further details regarding the error are documented here: https://github.com/armbian/build/pull/6057

Several people already tested the commit on their devices. I also tested it on two of my R4S (one Enterprise Edition, one Standard Edition). For me nothing changes, but I also did not face the issue before, so more test feedback is highly welcomed.

It would be great if this could be merged into OpenWrt. Another user reported it upstream (https://lore.kernel.org/linux-rockchip/CAG-UM5EbD7eJPO7+rRnZsV4mPfeFbtGmz18+23P1ZJgdqpFdAg@mail.gmail.com/T/#u), but it could take quite some time until this gets included in the kernel. Until then, some R4S users could struggle, that's why I'd highly appreciate if this could be included in OpenWrt for now.

Update 1: I originally did not want to mix patches, but decided to also added this patch, since it is related to PCI LAN reboot issues ... https://github.com/armbian/build/pull/7572

Djfe commented:
This is huge, thank you for looking into this issue.
We were experiencing issues on reboots but couldn't yet figure out why. this might explain our issues.

Ramon00 commented:
Please merge indeed. 
P.S. For reference this is linked to [issue 17923](https://github.com/openwrt/openwrt/issues/17923)

Djfe commented:
We observed a similar/the same issue on warm boots for the NanoPi R2S.
I hope/bet this will fix the rk3328 and others, too.

wurzerj commented:
Yes, this problem is absolutely strange since only some devices are affected. I own three R4S and could never reproduce the issue + then mostly blamed it on SD cards and power supplies (which sometimes cause issues).

@Djfe: It would be terrific if you could also test this, hoping that you have access to affected devices. Not sure if you have merge permissions… the formality checks are false positives from my perspective.

@Ramon00: I believe you tested this on affected devices already, right? Many thanks … in essence you did most of the heavy lifting.

Ramon00 commented:
I did not test yet, i dont have a build environment set up, and it is my main router, so i cannot mess around with it too much. But it appears to have been tested on armbian.

T0biii commented:
i may got a affected device but only experience this between firmware updates. 
I can only test if we build new gluon firmware for this openwrt patch and see if my device survives the update

wurzerj commented:
Did anyone get their hands on an affected device and can confirm whether this fixes the problem? As said, I own three R4S, but I never witnessed this problem myself. Nevertheless I am running a custom built including this patch and so far it does not harm.

T0biii commented:
> Did anyone get their hands on an affected device and can confirm whether this fixes the problem? As said, I own three R4S, but I never witnessed this problem myself. Nevertheless I am running a custom built including this patch and so far it does not harm.

Unfortunately I haven't been able to test it yet, but I hope to be able to next week, but I was already able to build test firmware for Gluon from Freifunk Munich with this patch

What @Djfe also noticed was the following comment: https://github.com/armbian/build/issues/6655#issuecomment-2538489984
maybe it would be good to implement that too?

teqlee commented:
Path had fixed issue of my R4S standard edition.

hvh-blet commented:
patch working! nanopi r4s 4gb standard edition

hauke commented on target/linux/rockchip/patches-6.6/401-rk3399-fix-pci-phy-reset-on-probe.patch:
```diff
@@ -0,0 +1,47 @@
++					   PHY_LANE_IDLE_MASK,
++					   PHY_LANE_IDLE_A_SHIFT + i));
++
++	reset_control_assert(rk_phy->phy_rst);
```
Please check the return value.

wurzerj replied:
Hi Hauke! Thanks for taking a look at this. Sorry, but could you explain what you mean? This follows a similar pattern than other functions in phy-rockchip-pcie.c ... for example line 263 (https://github.com/torvalds/linux/blob/v6.6/drivers/phy/rockchip/phy-rockchip-pcie.c).

hauke replied:
The other code paths use something like this:
```
	err = reset_control_assert(rk_phy->phy_rst);
	if (err)
		dev_err(&phy->dev, "assert phy_rst err %d\n", err);
	return err;
```

hauke commented:
Please extend the description of the patch a bit and send it to the upstream Linux mailing list too.

wurzerj reviewed (commented):
I believe the return value looks ok. It follows a similar pattern than other code in the existing phy-rockchip-pcie.c file.

Djfe commented:
this is not a regression. v2023.2 is affected as well.
This bug was always there

wurzerj commented:
> > Please also fix the problems the OpenWrt CI complains about.
> 
> @wurzerj Are you able to clear the failing checks?
> 
> Apologies if this comes across as obnoxious - not my intent at all. Just want to make sure you didn't miss this. I'm extremely grateful that you are pushing this along as I maintain a remote R4S and have been actively monitoring when this PR is accepted. I do not want to risk upgrading the R4S to 24.10 until this 24.10 regression is fixed.

Yes, I’m just terribly busy at the moment. Hopefully I’ll find some time on the weekend.

Ramon00 commented:
> > this is not a regression. v2023.2 is affected as well.
> > This bug was always there
> 
> Thank you for clarifying that. I won't delay upgrade to 24.10 now.

On 23.05 I do not have an issue, but I do on 24.10. Your mileage may vary

wurzerj commented:
Quick update: I have an updated version of this almost done (addressing haukes feedback) and also included https://github.com/armbian/build/pull/7572. This should hopefully holistically address the reboot issues occurring on some devices. I'll update the pull request (likely later today) after creating a build and testing it briefly.

wurzerj commented:
@T0biii: I now also included the other patch, hoping that this finally resolves all reboot issues.

@hauke: Patches are updated, I hope the formality checks now run smoothly (still queued). Regarding your return value comment. I believe it is best to not change the code. The script uses the plain "reset_control_assert(rk_phy->phy_rst);" code at least once. The newly introduced function does not need to return anything, but merely trigger the reset. Changing it would just increase complexity from my perspective. I'll submit upstream as soon as possible.

wurzerj committed:
rockchip: rk3399 fix PCIE PHY reset on probe

This patches the PCIE initialization error after warm reboot. The root of cause is, when the device is booted after power on, PHY stays in 'factory' state. After warm boot PHY stays in the previous state and prevents any training, thus PCIE init fails. Patch forked from Armbian (for further details see armbian/build#6057).

Signed-off-by: Juergen Wurzer <wurzer.juergen@gmail.com>

wurzerj committed:
rockchip: rk3399 disable PHY_LANE_IDLE_OFF for each instance of rockchip_pcie_phy_power_one

This patch fixes an issue in the Rockchip PCIe PHY driver where, after a warm restart of the rockchip_pcie_phy module, PCIe lanes other than lane 0 could remain stuck in the PHY_LANE_IDLE_OFF state. This resulted in the PCIe link being restricted to x1 mode, even in configurations designed to use multiple lanes. Patch forked from Armbian (for further details see armbian/build#7572).

Signed-off-by: Juergen Wurzer <wurzer.juergen@gmail.com>

hauke commented:
I do not think it gets more complex when you just add a logging when the reset failed.

Please post the link to the mailing list post when you send it upstream. 

@1715173329 Are you aware of this PR?

1715173329 commented:
tbh I don't know about PCIe stuffs and I don't have RK3399 boards now so it would be better to be reviewed by upstream.

Ramon00 commented:
any outlook on getting this merged before 24.10.1? Im kinda stuck on 23.05 right now. Thank you

hvh-blet commented:
OpenWrt 24.10.1 is out — still no fix!
LAN keeps dropping after reboot, as usual.

1715173329 commented:
Someone says this will break nvme from working

hvh-blet commented:
> Someone says this will break nvme from working

Which NVMe in the R4S???
Where exactly is it in the R4S? And what now, just leave it broken?!
Everything works fine in 23.05.5, but in 24 — it doesn't!
I would’ve stayed on 23.05.5, but the WiFi USB drivers are only available for 24.

1715173329 commented:
Some other rk3399 boards have NVMe slot.

hvh-blet commented:
> Some other rk3399 boards have NVMe slot.

So what now? Just forget about the R4S?
It's broken — and let it stay broken, is that it?!
Why does everything work perfectly on 23.05.5, but on 24 it's suddenly an issue?
Oh, I see — fixing it might "break something for someone else," right?
What kind of logic is that?!

1715173329 commented:
If a "fix" introduces new regressions, then it's just another bug, not a fix.
For affected users, what you said is also what they want to say to you.
Shouting here won't help anything, instead, just make things even worse.

hvh-blet commented:
> If a "fix" introduces new regressions, then it's just another bug, not a fix. For affected users, what you said is also what they want to say to you. Shouting here won't help anything, instead, just make things even worse.

First of all — you don’t even know what it sounds like when I’m really shouting.
Second — I bought this router because I want to use it at 100% of its potential.
But because of this bug in openwrt 24, I’m forced to stick with 23.05.5 and use another router as an access point,
since the drivers are only available in openwrt 24 — and openwrt 24 is just one big buggy mess.

Djfe commented:
@hvh-blet chill. I kindly ask you to stop polluting this PR.

I get it, you expected this to be fixed faster but please respect that developers are trying to solve this thoroughly. Also Noone is paid for their work here, least of all by you. This is not a paid distro like Red Hat so don't throw these messages at us.
Thank you.

Noone gave any promise when this would be fixed. And people are working on this in their free time so let them handle it at their pace. This is how stuff works around here and always has.
The NanoPis had several kinds of issues from the beginning that had to be ironed out by people in the Linux kernel and in OpenWrt. Issues that you might not even know about. Be kinder to them

PikeoftheLake commented:
I was messing with my R4S all afternoon after trying to upgrade to Openwrt 24.

I would constantly get no lan coming up on boot. Openwrt 23 showed no problem.

Interestingly enough, this seems to work great: https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds/releases

Any thoughts on why anaelorlinski's build works but the vanilla build doesn't?

hvh-blet commented:
> @hvh-blet chill. I kindly ask you to stop polluting this PR.
> 
> I get it, you expected this to be fixed faster but please respect that developers are trying to solve this thoroughly. Also Noone is paid for their work here, least of all by you. This is not a paid distro like Red Hat so don't throw these messages at us. Thank you.
> 
> Noone gave any promise when this would be fixed. And people are working on this in their free time so let them handle it at their pace. This is how stuff works around here and always has. The NanoPis had several kinds of issues from the beginning that had to be ironed out by people in the Linux kernel and in OpenWrt. Issues that you might not even know about. Be kinder to them

сосни хуйца, и завали ебло! забыл тебя спросить

hvh-blet commented:
> I was messing with my R4S all afternoon after trying to upgrade to Openwrt 24.
> 
> I would constantly get no lan coming up on boot. Openwrt 23 showed no problem.
> 
> Interestingly enough, this seems to work great: https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds/releases
> 
> Any thoughts on why anaelorlinski's build works but the vanilla build doesn't?

The vanilla version didn’t include the fix that corrects the LAN port operation! Yet another useless update. Buying the R4S was money down the drain!

Djfe commented:
> I was messing with my R4S all afternoon after trying to upgrade to Openwrt 24.
> 
> I would constantly get no lan coming up on boot. Openwrt 23 showed no problem.
> 
> Interestingly enough, this seems to work great: https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds/releases
> 
> Any thoughts on why anaelorlinski's build works but the vanilla build doesn't?

The changelog of the 2nd release in that list _OpenWrt 24.10 2025.03.23_  comes with the message
> Cherry pick some ImmortalWRT patches for Rockchip target

This includes the patches from this PR from the armbian repo
https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds/blob/main/openwrt-24.10/patches/target/linux/rockchip/patches-6.6/402-1-rk3399-fix-pci-phy-reset-on-probe.patch

https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds/blob/main/openwrt-24.10/patches/target/linux/rockchip/patches-6.6/402-2-rk3399-fix-pci-lanes.patch

1715173329 commented:
> Cherry pick some ImmortalWRT patches for Rockchip target

I guess this statement actually refers to patch 501 604 610.

Djfe commented:
ah that might be true. It's in the same commit and the same release tag though so I might've assumed wrong here
https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds/commit/6bc5ce1a7e0452412a03db594879138a5e47c042

davidandreoletti commented:
@wurzerj I have a device affected by this issue. I can offer to build + test this PR once 'ready'. 

Is https://github.com/openwrt/openwrt/pull/18078#issuecomment-2808480542 helping you decide what's the next course of action ?

hauke commented:
Please send this to the upstream Linux pcie mailing list too and work with them to get this merged upstream. 
I do not want to maintain such a patch in Openwrt for ever.

You still haven't implemented and tested my suggestions from https://github.com/openwrt/openwrt/pull/18078#discussion_r1984083643

eocrawford commented:
FWIW, I rolled a private build of 24.10.0 with the patches in this PR and can confirm that they fix the reboot issue on my affected R4S.

davidandreoletti commented:
> FWIW, I rolled a private build of 24.10.0 with the patches in this PR and can confirm that they fix the reboot issue on my affected R4S.

@eocrawford I understand you compiled the image yourself. If not a data sensitive image build, I would like run it on my end too and give a feedback too. Would you mind sharing it ?

eocrawford commented:
> > FWIW, I rolled a private build of 24.10.0 with the patches in this PR and can confirm that they fix the reboot issue on my affected R4S.
> 
> @eocrawford I understand you compiled the image yourself. If not a data sensitive image build, I would like run it on my end too and give a feedback too. Would you mind sharing it ?

Sure. Not sure if there's a better way to share images, but I put them on OneDrive, see: https://1drv.ms/f/c/4a339fb6fa8716cb/EpoWYJR8-UZIoiRJNwBy4PsB4bqD5n7NLXgcG-CCiioHbA?e=iAMc34. I built these about 5 days ago from a private branch based on git tag '24.10.0' that includes the two files in this PR.

davidandreoletti commented:
> > > FWIW, I rolled a private build of 24.10.0 with the patches in this PR and can confirm that they fix the reboot issue on my affected R4S.
> > 
> > 
> > @eocrawford I understand you compiled the image yourself. If not a data sensitive image build, I would like run it on my end too and give a feedback too. Would you mind sharing it ?
> 
> Sure. Not sure if there's a better way to share images, but I put them on OneDrive, see: https://1drv.ms/f/c/4a339fb6fa8716cb/EpoWYJR8-UZIoiRJNwBy4PsB4bqD5n7NLXgcG-CCiioHbA?e=iAMc34. I built these about 5 days ago from a private branch based on git tag '24.10.0' that includes the two files in this PR.

Thanks. I will test on a NanoPi R4S 4G (non enterprise).

syntron commented:
It seems to be that I also have a device affected by this - it is working with FriendlyWRT, not working with openwrt 24.10 but again working with anaelorlinski build mentioned above. If any testing is needed, I'm willing to try ;-)

davidandreoletti commented:
> ... not working with openwrt 24.10 ...

@syntron

 Is @eocrawford's [image build](https://1drv.ms/f/c/4a339fb6fa8716cb/EpoWYJR8-UZIoiRJNwBy4PsB4bqD5n7NLXgcG-CCiioHbA?e=iAMc34) resolving this issue on your hardware ? Please try it and report back.

syntron commented:
> > It seems to be that I also have a device affected by this - it is working with FriendlyWRT, not working with openwrt 24.10 but again working with anaelorlinski build mentioned above. If any testing is needed, I'm willing to try ;-)
> 
> > ... not working with openwrt 24.10 ...
> 
> @syntron Unclear if "not working with openwrt 24.10" refers to @eocrawford's [image build 24.10](https://1drv.ms/f/c/4a339fb6fa8716cb/EpoWYJR8-UZIoiRJNwBy4PsB4bqD5n7NLXgcG-CCiioHbA?e=iAMc34) or the official OpenWRT image.
> 
> Is @eocrawford 's image build resolving the issue on your hardware ?

Sorry if I was not specific - it is the official build which is not working; I will check @eocrawford's image now (if I can download it - was not workinbg on my first trys)

syntron commented:
> Sorry if I was not specific - it is the official build which is not working; I will check @eocrawford's image now (if I can download it - was not workinbg on my first trys)

I can confirm that @eocrawford's image is working well!

davidandreoletti commented:
The core issue seems fixed and verified by 3 independent individuals.

What's left to move this PR forward ?

1715173329 commented:
This patchset should be sent and reviewed by upstream kernel. There're also some reports that it will break NVMe but I got no log.

dmsza commented:
I did a 24.10.1 build and included this PR. I am still seeing the "stuck on reboot" issue in the R4S. The first reboot after a power on works, but the second reboot fails.

davidandreoletti commented:
@dmsza How do you differenciate older vs newer hardware ? (Eg: manufacturing date?)

dmsza commented:
@davidandreoletti I don't have a way to differentiate the hardware, and AFAIK they should be identical. To avoid confusion, I edited my post above and I replaced "older" and "newer" by the purchase date (May/2022 and Jan/2023 respectively). When I have time, I will open both units and take a picture of the PCB to check if there is any marking or other information that could confirm (or not) if they are the same hardware.

What puzzles me is the fact that this issue only start happening with 24.10.x (23.05.5 build works fine in both units).

davidandreoletti commented:
> When I have time, I will open both units and take a picture of the PCB to check if there is any marking or other information that could confirm

Thank you.

Ramon00 commented:
See https://www.eevblog.com/forum/repair/help-identifying-component-sot23-6-(hscu)/

dmsza commented:
Found it. It seems that this "HSCU" is a serial EEPROM used to store the MAC address. So the previous unit I purchased had that option. Probably it means I need to use the 24.10.x R4S Enterprise builds on this unit.

Still, going back on topic: this PR did not solve the reboot issue with 24.10.x.

![R4S_MACchip](https://github.com/user-attachments/assets/afcf809e-e376-42c8-aead-128cbb5c3bac)

davidandreoletti commented:
Friendelec mentioned the ship  24AA025E48T (you named it HSCU) is available on the NanoPi R4S Enterprise edition ONLY and is responsible for provide a global unique MAC address.

https://wiki.friendlyelec.com/wiki/index.php/NanoPi_R4S#Differences_Between_R4S_Standard_Version_.26_R4S_Enterprise_Version

As per your previous message and without further info so far, the fix is suitable for ... NanoPi R4S standard edition (4G) and corrobores with my experience + someone else here on the NanoPi R4S standard (4G).

IMHO, the fix resolves the issue entirely for "standard edition (4G )" users and should be upstreamed.

dmsza commented:
> @dmsza It seems it is not the only variable in play.
> 
> Friendelec mentioned the ship 24AA025E48T (you named it HSCU) is available on the NanoPi R4S Enterprise edition ONLY and is responsible for provide a global unique MAC address.
> 
> https://wiki.friendlyelec.com/wiki/index.php/NanoPi_R4S#Differences_Between_R4S_Standard_Version_.26_R4S_Enterprise_Version
> 
> I have a "standard edition (4G)" (ie no HSCU) and this PR fixed the lan interface issue on my device and someone else.
> 
> Do you a motherboard revision number on your devices ?

Well, originally when I purchased the unit with the 24AA025E48T chip FriendlyElec was not using the "Enterprise" name. 

But anyway, notice that I have two R4S units, read my post above, I believe it is clear that this PR does fix the issue with the R4S "Enterprise" LAN port not coming up with 24.10.

Also notice that my second R4S unit (regular, without the 24AA025E48T chip) does not have the LAN port issue, so LAN port comes up normally with or without this PR with 24.10. Also notice that only this unit has the reboot freeze issue with 24.10, which this PR does not make any difference.

These are my observations, I am in favor of upstreaming this PR. I just wanted to share that this PR will not fix the reboot freeze for some regular units without the 24AA025E48T chip.

nshopik commented:
Just wondering if I hit same issue as well. I've seen random speed renegotiaion on LAN port for my R4S.
` Generic FE-GE Realtek PHY r8169-0-100:00: Downshift occurred from negotiated speed 1Gbps to actual speed 100Mbps, check cabling!`. 

I've not seeing any boot issues with lan port though.

davidandreoletti commented:
> Downshift occurred from negotiated speed

What this happening on OpenWRT 23.x too (with presumably the same hardware on the network) ?

nshopik commented:
Yeah actually I never tried to downgrade, will try do that and report back.

dmsza commented:
Please apologize for the long comment, but I believe the information below may be helpful for the OpenWrt developers.

OK, last night I took some time do to some additional tests with the **two R4S units** I have. I tested with OpenWrt **22.03.7**, **23.05.5** and **24.10.1** (official builds, images downloaded from https://downloads.openwrt.org/).

The issues I am experiencing with 24.10.x is as follows:

- **Problem (A)**: In Unit 1, when powering on, **LAN does not come up**. This PR fixes this issue.
- **Problem (B)**: In Unit 2, after a cold start, **reboot hangs** every time. A power cycle (power disconnect/reconnect) is required to make it work again.

I discovered that **both issues are regressions**. Problem (B) is a well know problem as [explained in depth in this article](https://kohlschuetter.github.io/blog/posts/2022/10/28/linux-nanopi-r4s/) - voltage regulator configuration. I believed it was solved in 22.03.7 (since it does not have this problem), but reappeared in 23.05.7.

Below is a summary of my tests:

| Test  | 22.03.7 | 23.05.5 | 24.10.1 |
| ------------- | ------------- | ------------- | ------------- |
| (A) LAN enabled on power-on (cold start) |  Unit 1: OK<br>Unit 2: OK |  Unit 1: OK<br>Unit 2: OK |  Unit 1: **FAIL**<br>Unit 2: OK |
| (B) Reboot works after power-on (warm restart)   |  Unit 1: OK<br>Unit 2: OK |  Unit 1: OK<br>Unit 2: **FAIL** (*) |  Unit 1: OK<br>Unit 2: **FAIL** (*) |

_(*) I tested with at least six different SD cards from various brands (SanDisk, WD, Kingston, HP). Reboot only worked reliably with the old HP card—sometimes it takes longer to reboot, but it always works. I believe this is related to the [article about this issue](url) that says that non-UHS SD cards should work._

Djfe commented:
the last link is defect

PikeoftheLake commented:
Has anyone tried 24.10.2 released today?

zifirk commented:
> Has anyone tried 24.10.2 released today?

I am also interested in this question! I sold my r4s the other day. and I'm thinking of buying a new one.

SuPeR-NoVa-UK commented:
I'm also one of many whos R4S loses Lan after a reboot or firmware update, However FriendlyWRT dose not have this issue? It's just in a box until a release has been released to fix the issue

zifirk commented:
> > Has anyone tried 24.10.2 released today?
> 
> Yes I tested it today first worked and worked also afrer 1st reboot restored my old settings, but now lan does not work even after power cycle

Just sell it and stop wasting your nerves! I sold mine and went for an R3S. I’ll get another R4S once they finally fix this issue… but that doesn’t look like it’s happening anytime soon.

timfeierabend commented:
I have a standard NanoPI R4S 4GB (no MAC chip (enterprise edition), no EMMC), purchased on 12/4/24 via Amazon. I've never been able to get an official 24.10 release to work correctly - the LAN port doesn't start up correctly.

I compiled 2 versions of 24.10.2 - with and without the patches in this pull request. I used the default NanoPI R4S device target options + LuCI collection selected using `make menuconfig`. The unpatched build boots up w/ functioning LAN and WAN ports, but after a software reboot, the LAN port doesn't come back up. The patched version version works correctly, even after a software reboot.

cryobry commented:
The [upstream patch submission](https://lore.kernel.org/linux-rockchip/CAG-UM5EbD7eJPO7+rRnZsV4mPfeFbtGmz18+23P1ZJgdqpFdAg@mail.gmail.com/T/#u) mentioned in the OP was not done correctly. This fix will never get any traction three unless it is submitted correctly with inline patches to the mailing list, preferably by the original author.

abysso2 commented:
If anyone can tell me how - i will do it ...

cryobry commented:
> If anyone can tell me how - i will do it ...

```bash
git config --global user.name "Your Name" && \
git config --global user.email "you@example.com" && \
git config --global sendemail.smtpserver smtp.example.com && \
git config --global sendemail.smtpuser you@example.com && \
git config --global sendemail.smtpencryption tls && \
git config --global sendemail.smtpserverport 587 && \
./scripts/get_maintainer.pl -f drivers/pci/controller/pcie-rockchip.c && \
git commit --amend && \
git format-patch -1 HEAD && \
./scripts/checkpatch.pl 0001-*.patch && \
git send-email \
  --to=linux-rockchip@lists.infradead.org \
  --cc=linux-pci@vger.kernel.org \
  --cc=heiko@sntech.de \
  --cc=bhelgaas@google.com \
  0001-*.patch
```

abysso2 commented:
Hey @wurzerj , this is your work. Do you want to push this or shall i try?

wurzerj commented:
> Hey @wurzerj , this is your work. Do you want to push this or shall i try?

It would be amazing if you could give it a try. To be honest, I gave up and decided to not spend even more time on this. Sorry.

AntoineMazuyer commented:
@abysso2 hello Andreas, any news by chance ?

abysso2 commented:
Sorry for the late replay. Long story short: i did not succeed and will use the build of Anael (https://github.com/anaelorlinski/OpenWrt-NanoPi-R2S-R4S-Builds). The R4S is now a cold standby device. I switched to a OpenWRT One, because this is fully suported by the project.

PikeoftheLake commented:
Has anyone tried 24.10.3 to see if it works?  I could try tomorrow.

dmsza commented:
> I think it is time to close this PR

Allow me to disagree. Just like you, I replaced my two R4S units that I used as main routers. However, I do my own builds, and at least one of the R4S devices is working fine with my build and this PR applied. In fact, both R4S are working fine, but one still has the "hang on reboot" issue even with this PR applied.

That’s why I believe we should keep this PR open until these issues are fixed upstream and backported to OpenWrt. At least those who can do their own builds can apply this PR and keep their R4S devices running with the official OpenWrt.

wurzerj commented:
I thoroughly thought about this in the past couple of days and decided to close the pull request. There are a couple of reasons for this:
1) I do not see it as likely that this pull request will get merged into OpenWrt.
2) It seems it only partly fixed the issues based on the above feedback. For example @dmsza still seems to have a device not being able to reboot properly.
3) Despite closing this pull request one should still be able to apply it to custom OpenWrt builds.
4) I am not really able to further test this, since none of my 3 NanoPi R4S (two with unique mac, one without) ever showed those issues.
5) Despite really loving those R4Ss, I personally switched and now use two Radxa R52C, thus I'll not spend time with the R4S anymore. This also means that I'd not update this pull request, which would be necessary down the road.

I really hope the issue will get fixed upstream, since - as said - the R4S is a really great device.

wurzerj commented:
I thoroughly thought about this in the past couple of days and decided to close the pull request. There are a couple of reasons for this:
1) I do not see it as likely that this pull request will get merged into OpenWrt.
2) It seems it only partly fixed the issues based on the above feedback. For example @dmsza still seems to have a device not being able to reboot properly.
3) Despite closing this pull request one should still be able to apply it to custom OpenWrt builds.
4) I am not really able to further test this, since none of my 3 NanoPi R4S (two with unique mac, one without) ever showed those issues.
5) Despite really loving those R4Ss, I personally switched and now use two Radxa R52C, thus I'll not spend time with the R4S anymore. This also means that I'd not update this pull request, which would be necessary down the road.

I really hope the issue will get fixed upstream, since - as said - the R4S is a really great device.

1715173329 commented:
Has anyone tried this patchset?
https://lore.kernel.org/linux-rockchip/cover.1749582046.git.geraldogabriel@gmail.com/

dmsza commented:
@timfeierabend thank you very much. I have two R4S units - one affected by the reboot issue and the other by the LAN port issue. I noticed that this backport has already been committed to ```master``` (7ef19bb). If I have time, I will test it over the weekend and report back here.

dmsza commented:
> > the other unit, which did not have the LAN port issue but randomly hangs during soft reboots
> 
> try [this patch](https://github.com/immortalwrt/immortalwrt/blob/openwrt-24.10/target/linux/rockchip/patches-6.6/501-mmc-core-set-initial-signal-voltage-on-power-off.patch)

Thanks. Unfortunately it did not work. This issue is old, [this OpenWrt patch](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=2b583ab8a74a6bd41f1c9629b3b66ba561d6fceb) originally fixed this issue with kernel 5.10, and as I [reported above](https://github.com/openwrt/openwrt/pull/18078#issuecomment-2985030438), OpenWrt **22.03.7 does not have this issue**. This patch is [still included in the current 24.10](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/rockchip/patches-6.6/105-nanopi-r4s-sd-signalling.patch;h=b3c941821aaaf8a3474772df845acca401ac05fc;hb=refs/heads/openwrt-24.10) (kernel 6.6) builds, but for some reason it stopped working with kernels 5.15 (23.05) and later versions.

timfeierabend commented:
It would take a while to see which kernel changes between 5.10 and 5.15/6.6 could cause this issue. It looks like you're the only one on this PR reporting the hang on reboot issue - everyone else seems to be reporting issues with the LAN not working. Makes me think there is some kind of hardware variable going on with that one unit.

What kind of microsd card are you using? Can you provide an image of it? Are you using the same microsd card when testing both of your units?

Since you have access to the board, have you tried connecting a serial to usb debugger to the UART pins and pulled the kernel log messages?

PikeoftheLake commented:
> For everyone's awareness, the patch to fix the LAN reboot issue is now part of OpenWRT. The 24.10.4 release includes the LAN reboot fix discussed here. Everyone facing LAN port issues should be able to use mainline OpenWRT releases now.

Wonderful!  Many thanks for all the hard work with this.

davidandreoletti commented:
> For everyone's awareness, the patch to fix the LAN reboot issue is now part of OpenWRT. The 24.10.4 release includes the LAN reboot fix discussed here. Everyone facing LAN port issues should be able to use mainline OpenWRT releases now.

@timfeierabend Thank you for the wonderful effort!

dmsza commented:
I just tested ImmortalWrt 24.10.4 on the R4S unit that was hanging on soft reboot with OpenWrt (more details [here](https://github.com/openwrt/openwrt/pull/18078#issuecomment-2985030438)).

**ImmortalWrt does not have the issue** - the "problematic" unit reboots correctly every time. All other distros I tested (Armbian, DietPi, FriendlyWrt) also reboot without issues. Definitively, this problem is specific to OpenWrt official.

I noticed that ImmortalWrt’s `uboot-rockchip` package has some significant differences compared to OpenWrt’s. I will try building official OpenWrt and replacing only the `uboot-rockchip` package with the one from ImmortalWrt.
